RecoLocalCalo/HGCalRecProducers/plugins/HGCalCLUEAlgo.cc

```diff
@@ -216,89 +231,253 @@ math::XYZPoint HGCalCLUEAlgo::calculatePosition(const std::vector<int>& v, const
-        int binId = lt.getGlobalBinByBin(xBin, yBin);
-        size_t binSize = lt[binId].size();
-
-        for (unsigned int j = 0; j < binSize; j++) {
-          unsigned int otherId = lt[binId][j];
-          if (distance(i, otherId, layerId) < delta_c) {
-            cellsOnLayer.rho[i] += (i == otherId ? 1.f : 0.5f) * cellsOnLayer.weight[otherId];
+    bool isSi;
+    if (isOnlySi) {
+      isSi = true;
+    } else if (cellsOnLayer.isSi[i]) {
+      isSi = true;
+    } else {
+      isSi = false;
+    }
+    float delta(0);
+    if (isSi) {
+      delta = delta_c;
+      std::array<int, 4> search_box = lt.searchBox(
+          cellsOnLayer.x[i] - delta, cellsOnLayer.x[i] + delta, cellsOnLayer.y[i] - delta, cellsOnLayer.y[i] + delta);
+
+      for (int xBin = search_box[0]; xBin < search_box[1] + 1; ++xBin) {
+        for (int yBin = search_box[2]; yBin < search_box[3] + 1; ++yBin) {
+          int binId = lt.getGlobalBinByBin(xBin, yBin);
+          size_t binSize = lt[binId].size();
+
+          for (unsigned int j = 0; j < binSize; j++) {
+            unsigned int otherId = lt[binId][j];
+            bool otherSi;
+            if (isOnlySi) {
+              otherSi = true;
+            } else if (cellsOnLayer.isSi[otherId]) {
+              otherSi = true;
+            } else {
+              otherSi = false;
+            }
+            if (otherSi) {  //silicon cells cannot talk to scintillator cells
+              if (distance(i, otherId, layerId, false) < delta) {
+                cellsOnLayer.rho[i] += (i == otherId ? 1.f : 0.5f) * cellsOnLayer.weight[otherId];
+              }
+            }
           }
         }
       }
+    } else {
+      delta = delta_r;
+      std::array<int, 4> search_box = lt.searchBoxEtaPhi(cellsOnLayer.eta[i] - delta,
+                                                         cellsOnLayer.eta[i] + delta,
+                                                         cellsOnLayer.phi[i] - delta,
+                                                         cellsOnLayer.phi[i] + delta);
+      cellsOnLayer.rho[i] += cellsOnLayer.weight[i];
+      float northeast(0), northwest(0), southeast(0), southwest(0), all(0);
+
+      for (int etaBin = search_box[0]; etaBin < search_box[1] + 1; ++etaBin) {
+        for (int phiBin = search_box[2]; phiBin < search_box[3] + 1; ++phiBin) {
+          int binId = lt.getGlobalBinByBinEtaPhi(etaBin, phiBin);
+          size_t binSize = lt[binId].size();
+
+          for (unsigned int j = 0; j < binSize; j++) {
+            unsigned int otherId = lt[binId][j];
+            if (!cellsOnLayer.isSi[otherId]) {  //scintillator cells cannot talk to silicon cells
+              if (distance(i, otherId, layerId, true) < delta) {
+                // need to round these numbers to 5 decimals to be able to compare floats
+                constexpr float roundDec = 100000.;
+                float iPhi = round(cellsOnLayer.phi[i] * roundDec) / roundDec;
```

Comment:
curiously enough, between here and 10 lines below I lost track that this is a float. (just from the overwhelming ieta iphi in ints in other places in CMSSW)

Reply:
What's the significance of the 5th decimal in this game? Is it important to be equal in this epsilon range of differences?
Where is the problem with simply using floats as is?

Reply:
I have almost done changing the comparison using iEta and iPhi.